### PR TITLE
[c10d] Convenience wrappers for collective functions

### PR DESCRIPTION
**Pull request by pietern**

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -71,10 +87,31 @@ PyObject* c10d_init(PyObject* _unused) {
...
+          .def(
+              "broadcast",
+              [](::c10d::ProcessGroup& pg, at::Tensor& x, int rootRank) {
+                ::c10d::BroadcastOptions opts;
+                opts.rootRank = rootRank;
+                std::vector<at::Tensor> xs = {x};
+                return pg.broadcast(xs, opts);
+              },
+              py::arg("tensor"),
+              py::arg("root"))
```
Comment (apaszke) [inline torch/csrc/distributed/c10d/init.cpp]: You forgot to release the GIL here

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -71,10 +87,31 @@ PyObject* c10d_init(PyObject* _unused) {
...
+          .def(
+              "allreduce",
+              [](::c10d::ProcessGroup& pg, at::Tensor& x, ::c10d::ReduceOp op) {
+                ::c10d::AllreduceOptions opts;
+                opts.reduceOp = op;
+                std::vector<at::Tensor> xs = {x};
+                return pg.allreduce(xs, opts);
+              },
+              py::arg("tensor"),
+              py::arg("op") = ::c10d::ReduceOp::SUM);
```
Comment (apaszke) [inline torch/csrc/distributed/c10d/init.cpp]: Missing GIL release again

**Comment by pietern** [comment]
@apaszke Thanks, fixed.

**Review by apaszke — APPROVED** [review]

File: test/test_c10d.py [hunk 3]
```diff
@@ -136,6 +136,12 @@ def broadcast(xs, rootRank, rootTensor):
                 self.assertEqual(torch.Tensor([i * self.size + j]), xs[0])
                 self.assertEqual(torch.Tensor([i * self.size + j]), xs[1])
 
+        # Test overloaded convenience function
+        x = torch.Tensor([self.rank + 1.0])
+        work = pg.broadcast(x, root=0)
+        work.wait()
+        self.assertEqual(torch.Tensor([1.0]), x)
```
Comment (apaszke) [inline test/test_c10d.py]: Actually another improvement we could make (to make sure the nicer API is tested as thoroughly as the main one) would be to wrap the `for i` loop above in `for broadcast in [broadcast_using_options, broadcast_sugared]:` where `broadcast_*` functions are written like the one above

**Review by apaszke — APPROVED** [review]

**Pull request merged**